sourcegraph/cody
# Chat/Edit: Smart Apply

Closes CODY-3200

## Description

This PR:
- Adds a "Smart Apply" option to Chat. This feature:
  - Prompts the LLM to produce a selection in a file, and then prompts the LLM to apply the code block via inline edit
- Prompts the LLM to associate code blocks with file paths, via `language:path/to/file.ts` metadata, and parses that via a remark plugin
- Adds a simpler UI for codeblocks
  - Less hard to understand icons, more words
  - "Copy" and "Apply" are at the forefront
  - "Insert Code at Cursor" and "Save Code to New File" are behind a dropdown menu
  - The filename associated with the codeblock is shown	

### Demo

https://www.loom.com/share/70e3c47cb07e4a69b72515c4f214f76a?sid=5286f9d6-f25b-4d94-b936-8af5eb4f729e

### Screenshots

<img width="377" alt="image" src="https://github.com/user-attachments/assets/47e58738-1d76-4f25-b1d0-b2b91da65ef3">

<img width="429" alt="image" src="https://github.com/user-attachments/assets/d181d27a-ce98-42a0-984c-ed0519298952">


## Test plan

1. Enable the feature flag or `cody.experimental.unstable`
2. Check that the new UI appears
3. Check that the "Copy" button works, as well as "Insert Code at Cursor" and "Save Code to New File" work
4. Execute a smart apply on existing code in an expected existing file, check that it works
5. Execute a smart apply on new code in an expected existing file, check that it works
6. Execute a smart apply on new code in an expected new file, check that it works
